File: src/assets.go
package main

import (
	"bytes"
	"embed"
	"encoding/base64"
	"encoding/json"
	"path"
	"regexp"
	"strings"
)

// The Cloudflare Worker: routing and uploads in worker.js, the Room Durable
// Object in room.js. They are concatenated into one module at deploy time.
//
//go:embed worker/room.js worker/worker.js
var workerFS embed.FS

// The reader shell: real .html, .css and .js files, compiled into the Worker.
//
//go:embed shell
var shellFS embed.FS

// The wordmark's typeface, complete and unmodified, under the OFL text beside
// it. It is served from its own route rather than inlined, so it can be cached
// for a year while the stylesheet stays short-lived.
//
//go:embed shell/vendor/fonts/IBMPlexSans-SemiBold.woff2
var wordmarkFont []byte

// workerModules is ordered: room.js defines and exports the Room class, so
// putting it first makes worker.js's re-export resolve without a second module
// part in the upload.
var workerModules = []string{"room.js", "worker.js"}

// shellRoutes maps a request path to the file that answers it.
var shellRoutes = map[string]string{
	// Pico CSS, vendored under vendor/ so the shell keeps its no-build,
	// no-CDN property. See vendor/pico-LICENSE.md.
	"/pico.css":               "vendor/pico.css",
	"/index.html":             "index.html",
	"/reader.html":            "reader.html",
	"/404.html":               "404.html",
	"/komodoc.css":            "komodoc.css",
	"/reader.js":              "reader.js",
	"/agent.js":               "agent.js",
	"/anchor.js":              "anchor.js",
	"/documentation":          "documentation.html",
	"/assets/komodo-logo.svg": "assets/komodo-logo.svg",
	"/docs/commenting.png":    "assets/commenting.png",
	"/docs/sandbox.png":       "assets/sandbox.png",
}

var contentTypes = map[string]string{
	".html": "text/html; charset=utf-8",
	".css":  "text/css; charset=utf-8",
	".js":   "text/javascript; charset=utf-8",
	".png":  "image/png",
	".svg":  "image/svg+xml",
}

type shellFile struct {
	Type string `json:"type"`
	Body string `json:"body"`
	// Base64 says the body is an encoded binary rather than text. The Worker
	// carries the shell as JSON, which cannot hold arbitrary bytes, so a font
	// travels encoded and is decoded when it is served.
	Base64 bool `json:"base64,omitempty"`
	// Immutable marks a file whose bytes never change, so it can be cached
	// for a year instead of five minutes.
	Immutable bool `json:"immutable,omitempty"`
}

// fontRoute is served on its own, rather than inlined into the stylesheet,
// because a font never changes and a stylesheet does. Sharing one cache
// lifetime would have meant refetching the font whenever the CSS expired.
const fontRoute = "/fonts/ibm-plex-sans-600.woff2"

// The first heading of the rendered README, which the hero replaces.
var reLeadingTitle = regexp.MustCompile(`(?s)\A\s*<h1[^>]*>.*?</h1>`)

var reExport = regexp.MustCompile(`(?m)^export \{ Room \} from "\./room\.js";\n`)

// documentation renders the project README into the HTML the documentation
// page is built around. The README is the only copy of that text: it is
// embedded from the shell directory, where the build puts it, so the page
// cannot drift from what the repository says.
func documentation() string {
	source, err := shellFS.ReadFile("shell/README.md")
	if err != nil {
		die("missing README.md in the shell: run make build, which copies it in")
	}
	var body bytes.Buffer
	if err := markdown.Convert(source, &body); err != nil {
		die("could not render README.md: %v", err)
	}
	// The page opens with the mark and the wordmark, as the landing page does,
	// so the README's own title would say the name twice.
	return reLeadingTitle.ReplaceAllString(body.String(), "")
}

func loadShell() map[string]shellFile {
	settings, err := json.Marshal(config)
	if err != nil {
		die("could not encode the configuration: %v", err)
	}
	prose := documentation()
	shell := make(map[string]shellFile, len(shellRoutes))
	for route, name := range shellRoutes {
		body, err := shellFS.ReadFile("shell/" + name)
		if err != nil {
			die("missing shell file: %s", name)
		}
		if path.Ext(name) == ".png" {
			shell[route] = shellFile{
				Type:      contentTypes[".png"],
				Body:      base64.StdEncoding.EncodeToString(body),
				Base64:    true,
				Immutable: true,
			}
			continue
		}
		// The page checks the same size and file-type limits the server does,
		// so __CONFIG__ is substituted here too rather than duplicated in HTML.
		source := strings.ReplaceAll(string(body), "__CONFIG__", string(settings))
		// After the configuration, so nothing in the README is mistaken for a
		// placeholder the shell was meant to fill in.
		source = strings.ReplaceAll(source, "__README__", prose)
		shell[route] = shellFile{Type: contentTypes[path.Ext(name)], Body: source}
	}
	shell[fontRoute] = shellFile{
		Type:      "font/woff2",
		Body:      base64.StdEncoding.EncodeToString(wordmarkFont),
		Base64:    true,
		Immutable: true,
	}
	return shell
}

func loadWorker() string {
	parts := make([]string, 0, len(workerModules))
	for _, name := range workerModules {
		source, err := workerFS.ReadFile("worker/" + name)
		if err != nil {
			die("missing worker file: %s", name)
		}
		parts = append(parts, strings.Trim(reExport.ReplaceAllString(string(source), ""), "\n"))
	}
	return strings.Join(parts, "\n\n") + "\n"
}

// workerSource is the Worker module, with the reader shell and the shared
// limits from config.go compiled into it.
func workerSource() string {
	shell, err := json.Marshal(loadShell())
	if err != nil {
		die("could not encode the shell: %v", err)
	}
	settings, err := json.Marshal(config)
	if err != nil {
		die("could not encode the configuration: %v", err)
	}
	source := strings.Replace(loadWorker(), "__SHELL__", string(shell), 1)
	return strings.Replace(source, "__CONFIG__", string(settings), 1)
}

File: komodoc/src/assets.rs
//! The reader shell: the pages, the bundles they load, and the renderers they
//! render with, all compiled into the binary.
//!
//! What is embedded is a build output. `web/` holds the Svelte sources, bun
//! and vite build them into `src/shell`, and this serves whatever is there --
//! by path, rather than from a table of routes, because the names of a bundle's
//! files are decided by the bundler and change with their contents.

use std::collections::HashMap;

use include_dir::{include_dir, Dir, File};
use sha2::{Digest, Sha256};

use crate::config::Configuration;

static SHELL: Dir<'static> = include_dir!("$CARGO_MANIFEST_DIR/../src/shell");

/// The renderers, which are build outputs of the engine crate rather than of
/// the web build, and are addressed by a digest of their own bytes: a module
/// is fetched once per browser and cached for a year, so a fixed route that
/// outlived a rebuild would hand a stale module to a loader that no longer
/// speaks to it. The reader page is told the current URLs when it is served.
const MODULES: &[(&str, &str)] = &[
    ("markdown", "wasm/markdown.wasm"),
    // Optional: `make typst` builds it, and a build without it simply does not
    // list typst among its renderers.
    ("typst", "wasm/typst.wasm"),
];

/// The in-frame half of the reader, which the server writes into every
/// document it serves. It is built on its own and keeps a fixed name, because
/// that name is what goes into those documents.
#[allow(dead_code)]
pub const AGENT_ROUTE: &str = "/agent.js";

pub fn content_type(name: &str) -> &'static str {
    match name.rsplit('.').next() {
        Some("html") => "text/html; charset=utf-8",
        Some("css") => "text/css; charset=utf-8",
        Some("js") => "text/javascript; charset=utf-8",
        Some("json") => "application/json; charset=utf-8",
        Some("map") => "application/json; charset=utf-8",
        Some("png") => "image/png",
        Some("jpg") | Some("jpeg") => "image/jpeg",
        Some("svg") => "image/svg+xml",
        Some("wasm") => "application/wasm",
        Some("woff2") => "font/woff2",
        Some("txt") | Some("md") => "text/plain; charset=utf-8",
        _ => "application/octet-stream",
    }
}

#[derive(Clone, Debug)]
pub struct ShellFile {
    pub kind: &'static str,
    pub body: Vec<u8>,
    /// A file whose bytes never change under this name, so it can be cached
    /// for a year rather than five minutes. Everything the bundler names for
    /// its own contents is one, and so are the renderers.
    pub immutable: bool,
}

impl ShellFile {
    /// The file as text, for the pages that are text and the tests that read
    /// them.
    #[allow(dead_code)]
    pub fn text(&self) -> String {
        String::from_utf8_lossy(&self.body).to_string()
    }
}

fn file(name: &str) -> Option<&'static [u8]> {
    SHELL.get_file(name).map(File::contents)
}

/// Where a renderer is served from: its name under /wasm/, stamped with a
/// digest of the bytes, so the URL changes whenever the module does.
fn module_route(name: &str, body: &[u8]) -> String {
    let digest = hex::encode(Sha256::digest(body));
    format!("/wasm/{name}.{}.wasm", &digest[..16])
}

/// Where this build serves its renderers, for the tests that fetch one and
/// for the page that loads them.
#[allow(dead_code)]
pub fn module_url(name: &str) -> Option<String> {
    let path = MODULES.iter().find(|(known, _)| *known == name)?.1;
    Some(module_route(name, file(path)?))
}

/// The compiled typst renderer, if this build has one.
pub fn typst_module() -> Option<&'static [u8]> {
    file("wasm/typst.wasm")
}

/// The source formats this process can render in a reader, and so offer an
/// editor for. Markdown always; typst when the module was built.
pub fn renderers() -> Vec<String> {
    let mut list = vec!["markdown".to_string()];
    if typst_module().is_some() {
        list.push("typst".to_string());
    }
    list
}

/// The project README rendered into the HTML the documentation page is built
/// around. The README is the only copy of that text: it is embedded from the
/// shell directory, where the build puts it, so the page cannot drift from
/// what the repository says.
fn documentation() -> Result<String, String> {
    let source = file("README.md")
        .ok_or("missing README.md in the shell: run make build, which copies it in")?;
    let body = komodoc_engine::markdown::render_body(&String::from_utf8_lossy(source));
    // The page opens with the mark and the wordmark, as the landing page does,
    // so the README's own title would say the name twice.
    let leading = regex::Regex::new(r"(?s)^\s*<h1[^>]*>.*?</h1>").expect("a constant pattern");
    Ok(leading.replace(&body, "").to_string())
}

/// Everything the server answers with, ready to serve.
pub fn load_shell(_config: &Configuration) -> Result<HashMap<String, ShellFile>, String> {
    let prose = documentation()?;

    // The renderers first, because the reader page has to be told where they
    // are before it is served.
    let mut shell = HashMap::new();
    let mut modules = serde_json::Map::new();
    for (name, path) in MODULES {
        let Some(body) = file(path) else { continue };
        let route = module_route(name, body);
        modules.insert(name.to_string(), serde_json::Value::String(route.clone()));
        shell.insert(
            route,
            ShellFile {
                kind: "application/wasm",
                body: body.to_vec(),
                immutable: true,
            },
        );
    }
    let modules = serde_json::Value::Object(modules).to_string();

    for entry in walk(&SHELL) {
        let path = entry.path().to_string_lossy().to_string();
        // The renderers are served under their digest, and the README is the
        // documentation page's text rather than a page of its own.
        if path.starts_with("wasm/") || path == "README.md" {
            continue;
        }
        let kind = content_type(&path);
        let route = format!("/{path}");
        // The bundler names every asset for a digest of its own contents, so
        // one of those can be kept for a year; a page is rewritten in place by
        // the next build and cannot be.
        let immutable = path.starts_with("assets/") || path.starts_with("fonts/");
        let body = if kind.starts_with("text/html") {
            // The one thing a page is told when it is served rather than when
            // it is built: what this deployment holds. The prose of the
            // documentation page, and where the renderers are.
            entry
                .contents_utf8()
                .unwrap_or_default()
                .replace("__README__", &prose)
                .replace("__MODULES__", &modules)
                .into_bytes()
        } else {
            entry.contents().to_vec()
        };
        shell.insert(
            route,
            ShellFile {
                kind,
                body,
                immutable,
            },
        );
    }
    Ok(shell)
}

/// Every file in the shell, at any depth.
fn walk(dir: &'static Dir<'static>) -> Vec<&'static File<'static>> {
    let mut found: Vec<&File> = dir.files().collect();
    for child in dir.dirs() {
        found.extend(walk(child));
    }
    found
}
